File: java/src/main/java/it/giannibombelli/wsc2026/payment/application/query/PaymentFinder.java
package it.giannibombelli.wsc2026.payment.application.query;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.domain.primitive.ClientReference;
import it.giannibombelli.wsc2026.payment.domain.payment.Payment;
import it.giannibombelli.wsc2026.payment.domain.payment.PaymentId;
import it.giannibombelli.wsc2026.payment.domain.payment.Transaction;
import it.giannibombelli.wsc2026.payment.domain.ports.PaymentRepository;

import java.util.Optional;

import static java.util.Objects.requireNonNull;

public final class PaymentFinder {
    private final PaymentRepository paymentRepository;

    public PaymentFinder(PaymentRepository paymentRepository) {
        this.paymentRepository = requireNonNull(paymentRepository);
    }

    public Optional<PaymentDetails> findDetailsById(PaymentId id) {
        Require.requireArgument(id, "paymentId");

        return paymentRepository.findById(id).map(this::toDetails);
    }

    public Optional<PaymentSummary> findSummaryById(PaymentId id) {
        Require.requireArgument(id, "paymentId");

        return paymentRepository.findById(id).map(this::toSummary);
    }

    public Optional<PaymentDetails> findDetailsByClientReference(ClientReference clientReference) {
        Require.requireArgument(clientReference, "clientReference");

        return paymentRepository.findByClientReference(clientReference).map(this::toDetails);
    }

    private PaymentDetails toDetails(Payment payment) {
        return new PaymentDetails(
            payment.id().value(),
            payment.clientReference(),
            payment.amount(),
            payment.status(),
            payment.requestedAt(),
            payment.transactions().stream().map(this::toTransactionDetail).toList()
        );
    }

    private PaymentSummary toSummary(Payment payment) {
        return new PaymentSummary(
            payment.id().value(),
            payment.clientReference(),
            payment.amount(),
            payment.status(),
            payment.requestedAt()
        );
    }

    private PaymentDetails.TransactionDetail toTransactionDetail(Transaction transaction) {
        return new PaymentDetails.TransactionDetail(
            transaction.id().value(),
            transaction.provider(),
            transaction.providerReference() == null ? null : transaction.providerReference().value(),
            transaction.amount(),
            transaction.status(),
            transaction.startedAt(),
            transaction.completedAt()
        );
    }
}

File: typescript/src/payment/application/query/paymentFinder.ts
// Query service per recuperare i Payment in forma di read model.

import { ClientReference } from '@/common/domain/primitive/clientReference.js';
import { Payment } from '../../domain/payment/payment.js';
import { PaymentId } from '../../domain/payment/paymentId.js';
import { Transaction } from '../../domain/payment/transaction.js';
import type { PaymentRepository } from '../../domain/ports/paymentRepository.js';
import { PaymentDetails } from './paymentDetails.js';
import { PaymentSummary } from './paymentSummary.js';
import { requireArgument } from '@/common/utils/requireArgument.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class PaymentFinder {
  private readonly _paymentRepository: PaymentRepository;

  constructor(paymentRepository: PaymentRepository) {
    requireDependency(paymentRepository, "paymentRepository");
    this._paymentRepository = paymentRepository;
  }

  findDetailsById(id: PaymentId): PaymentDetails | null {
    requireArgument(id, 'paymentId');

    const payment = this._paymentRepository.findById(id);
    return payment === null ? null : this.toDetails(payment);
  }

  findSummaryById(id: PaymentId): PaymentSummary | null {
    requireArgument(id, 'paymentId');

    const payment = this._paymentRepository.findById(id);
    return payment === null ? null : this.toSummary(payment);
  }

  findSummaryByClientReference(clientReference: ClientReference): PaymentSummary | null {
    requireArgument(clientReference, 'clientReference');

    const payment = this._paymentRepository.findByClientReference(clientReference);
    return payment === null ? null : this.toSummary(payment);
  }

  findDetailsByClientReference(clientReference: ClientReference): PaymentDetails | null {
    requireArgument(clientReference, 'clientReference');

    const payment = this._paymentRepository.findByClientReference(clientReference);
    return payment === null ? null : this.toDetails(payment);
  }

  private toDetails(payment: Payment): PaymentDetails {
    return {
      id: payment.id().value.value,
      clientReference: payment.clientReference(),
      amount: payment.amount(),
      status: payment.status(),
      requestedAt: payment.requestedAt(),
      transactions: payment.transactions().map((t) => this.toTransactionDetail(t)),
    };
  }

  private toSummary(payment: Payment): PaymentSummary {
    return {
      id: payment.id().value.value,
      clientReference: payment.clientReference(),
      amount: payment.amount(),
      status: payment.status(),
      requestedAt: payment.requestedAt(),
    };
  }

  private toTransactionDetail(transaction: Transaction): PaymentDetails.TransactionDetail {
    const providerReference = transaction.providerReference();
    const completedAt = transaction.completedAt();
    return {
      id: transaction.id().value.value,
      provider: transaction.provider(),
      providerReference: providerReference === null ? null : providerReference.value.value,
      amount: transaction.amount(),
      status: transaction.status(),
      startedAt: transaction.startedAt(),
      completedAt: completedAt === null ? null : completedAt,
    };
  }
}
